### ISSUE TITLE
DeepFillv2 release

### ISSUE THREAD

**Issue by minmax100** [JiahuiYu/generative_inpainting#62]
Really nice work and great idea for the DeepFillv2!! Any plan and expected date to release the DeepFillv2 code?

**Comment by xhh232018** [comment]
I am quite amazed by the effect of DeepFillv2 !!! Also, I want to know the expected date of the code release since I want to know how it works. Thank you!

**Comment by yu45020** [comment]
+1
I try to swap all convolution with the gated one in the Mobile Net V2 and add another decoder for image in-painting. But the result is not good. 

Here is my PyTorch code for the gated convolution. I want to hear your feedback. 
```
# BaseModule simply adds more functions to PyTorch's base module. 
class PartialConvBlock(BaseModule):
    # mask is binary, 0 is masked point, 1 is not

    def __init__(self, in_channels, out_channels, kernel_size, stride=1,
                 padding=0, dilation=1, groups=1, bias=False, BN=True, activation=None):
        super(PartialConvBlock, self).__init__()
        self.conv = nn.Conv2d(in_channels, out_channels, kernel_size, stride,
                              padding, dilation, groups, bias)

        self.mask_conv = nn.Sequential(nn.Conv2d(in_channels, out_channels, kernel_size, stride,
                                                 padding, dilation, groups, bias))

        out = []
        if BN:
            out.append(nn.BatchNorm2d(out_channels))
        if activation:
            out.append(activation)
        if len(out):
            self.out = nn.Sequential(*out)

    def forward(self, args):
        x, mask = args
        feature = self.conv(x)
        mask = self.mask_conv(mask)
        x = feature * torch.sigmoid(mask)
        if hasattr(self, 'out'):
            x = self.out(x)
        return x, mask

```

**Comment by JiahuiYu** [comment]
First, thanks for your interest!
Currently we do NOT have an estimate date for code release. The only thing we know is that we will NOT release code in this summer. However, implementing deepfill v2 is relatively easy based on current code if you can understand all components of this repo.

@yu45020 It seems your code is not the gated convolution we proposed. Please check carefully of the tech report.
1. We do not have `x, mask = args`, the only input is `x`.
2. We do not have `BN` and I am not sure how it affects performance (In fact, in the earlier development of deepfill v1, we found `BN` slightly hurts color consistence. But if you can finetune model without BN in the end, it should be fine).
3. In your code, the default activation seems `None`. We use ReLU as activation.
4. Gated Conv can be implemented as:
```py
x1 = self.conv1(x)
x2 = self.conv2(x)
x = sigmoid(x2) * activation(x1)
```
**Or** a slightly faster implementation on GPU:
```
x = self.conv(x)
x1, x2 = split(x, 2) # split along channels 
x = sigmoid(x2) * activation(x1)
```

Hope it helps.

**Comment by yu45020** [comment]
Great! 
My code is used as a drop-in replacement for my current project, so there are BN and activation options. Sorry for the confusion. But I did misunderstand your paper on the mask.  Batch norm seems work great in black/white picture, but it seems hurt color pictures. 

By the way, I re-implement 'Image Inpainting for Irregular Holes Using Partial Convolutions' before coming to your paper, so I was wondering where should I put the mask. But I was wrong.

**Comment by limchaos** [comment]
Really simple and efficient way ! Great idea!
Hi, have you compared with vanilla D architecture without  Spectral Norm , I mean what's the benefit of SN in this work .

**Comment by JiahuiYu** [comment]
@limchaos Thanks for your interest first! You mentioned a good point: what's the performance difference w.r.t. different GAN architecture. The roadmap of GANs for inpainting is like: one vanilla D arch -> two global/local D arch -> SN-PatchGAN. We did not show comparison results in main paper because:
1. One vanilla D arch has worse performance, which is studied in [Globally and Locally Consistent Image Completion](https://github.com/satoshiiizuka/siggraph2017_inpainting).
2. GAN without SN has worse performance and is not stable, which is studied in [Spectral Normalization for Generative Adversarial Networks](https://openreview.net/forum?id=B1QRgziT-)
3. We do have ablation experiments and we may add it to appendix in future.

**Comment by liouxy** [comment]
@JiahuiYu Thanks for your excellent work. I'm tring to reimplement the DeepFillv2 based on your code. But the results seem not good enough, can i ask you some details about it?
1. In the paper, you mentioned that "our final objective function for inpainting network is only composed of pixel-wise reconstruction loss and SN-PatchGAN loss with default loss balancing hyper-parameter as 1:1". In the code v1 :
 losses['l1_loss'] = l1_alpha * tf.reduce_mean(tf.abs(local_patch_batch_pos - local_patch_x1)*spatial_discounting_mask(config)) 
losses['ae_loss'] = l1_alpha * tf.reduce_mean(tf.abs(batch_pos - x1) * (1.-mask))
losses['g_loss'] = config.GAN_LOSS_ALPHA * losses['g_loss']
 losses['g_loss'] += config.L1_LOSS_ALPHA * losses['l1_loss']
losses['g_loss'] += config.AE_LOSS_ALPHA * losses['ae_loss']
What is the detail form of the g_loss in DeepFillv2?

2. Since hinge loss is used to calculate the gan loss, what kind of activation function will be used at the last layer of discriminator?

3. Are there any neccesarry change in attention module since the mask is free-form?

Thank you again!

**Comment by JiahuiYu** [comment]
Hi, thanks for your interest.

1. Local patch loss is removed. It is AEloss + GANloss. L1 loss and AE loss is just same loss on different regions of image (inside mask or not). 
2. We use the setting of this repo: https://github.com/pfnet-research/sngan_projection/blob/master/updater.py
3. No. The contextual attention is already implemented to support any shape masks.

**Comment by JiahuiYu** [comment]
@liouxy Yes. foreground:background is 1:1 . Since the mask is no longer rectangular, it is difficult to calculate the discounted weights proposed in CVPR paper. So the final solution is just a simple pixel-wise loss over input and output.

**Comment by xhh232018** [comment]
@JiahuiYu Thanks for your former help. I've read your paper carefully for several times and the implementation is almost done. However. some details confused me so I need your help
1. In your paper, you mentioned that the input for discriminator is the predicted image, mask and the guidance(sketch) and the output is a 3-D shaped feature map with size of h*w*c. For the hinge loss calculation, can I directly put the 3-D feature map into the hinge loss formula?
2. The spectral normalization is implemented in your algorithm. However, I ignored it at first. I saw it but I made no response to it. Today, I spent some time reading the paper of SNGAN. I realized that the convolution layers of discriminant network are based on spectral normalization. Is it right? (I thought they were vanilla convolution layers 2 days ago. Obviously, I was wrong) Also, the leakyRelu and batch normalization are implemented in SNGAN. Should I add them to the discriminant network?

**Comment by JiahuiYu** [comment]
@xhh232018 @yu45020  I recommend this official implementation of [SN-GAN](https://github.com/pfnet-research/sngan_projection/blob/master/updater.py#L21).

**Comment by xhh232018** [comment]
@yu45020 I got the same problem of GAN training the g_loss is quite high. Have you solved this problem?

**Comment by yu45020** [comment]
@xhh232018 
Not yet. I follow the structure in [deep fill](https://github.com/JiahuiYu/generative_inpainting/blob/06cd62cfca8c10c349b451fa33d9cbb786bfaa20/inpaint_model.py#L29) and swap all convolutions with gated convolution. The GAN part is similar to the paper's part. The loss stays high. 

I find L1 loss decreases quickly but hinge loss decreases slowly.

Here are what I have tried but don't work: 
*  change the convolutions ( 5x5)  in GAN to separable convolutions (like EffNet) in order to increase batch size
* add batch norm in the generator 
* concatenate decoder with corresponding encoder

I have not trained the model for a long time since I rent GPU.  I am wondering whether using pre-trained networks as discriminator will help. 

But if I replace the partial convolution with gated convolution and use the complicated loss function from [this paper](https://arxiv.org/abs/1804.07723), the improvement is visible within few iterations.  I might misunderstand some details in GAN.

**Comment by xhh232018** [comment]
@yu45020  An expert told me that It is common that training GAN is a quite long process. Also the author mentioned that there is no batch-norm operation in the network. I have trained the Deepfill V1 for 5 days in order to get good results.  I tried PCONV  2 weeks ago. It converges very quickly because its structure is relatively simple. I only trained the Deepfill V2 for 2 days. I will see what it will happen after several days training..

**Comment by yu45020** [comment]
Thanks for the info. I notice the deep fill v2 removes batch norm to reduce color consistence since the author mentioned it above. But my application focuses on black/white images, and batch norm helps.

**Comment by shihaobai** [comment]
![image](https://user-images.githubusercontent.com/42648726/44540528-aa7b9380-a739-11e8-81f5-3e76d7959b97.png) @JiahuiYu  what's the meaning of the input x? Does it refer to the incomplete image?  Is it the same as what defined in v1 code?  In the code v1: x = tf.concat([x, ones_x, ones_x*mask], axis=3)

**Comment by JiahuiYu** [comment]
@shihaobai You are correct. `x` means input incomplete image, which is exactly `x = tf.concat([x, ones_x, ones_x*mask], axis=3)`.

**Comment by shihaobai** [comment]
@JiahuiYu  Thanks for your reply. It does help. I also have an another question.  I tried to use relu as activation for gated-conv. But i found my d_loss reached convergence quickly so that my generator couldn't learn well. When i used elu instead of it. The model worked better. So i think there must be something wrong in my discriminator or the frequency i trained my generator and discriminator. How often do you train the discriminator and generator? and here is my discriminator code:
`def dis_conv(x, cnum, ksize=5, stride=2, name='conv', training=True):
     x_shape=x.get_shape().as_list()
    w=tf.get_variable(name=name+'_w',shape=[ksize,ksize,x_shape[-1]]+[cnum])
    w=spectral_norm(w,name=name)   
    x = tf.nn.conv2d(x,w,strides=[1,stride,stride,1],padding='SAME')
    bias=tf.get_variable(name=name+'_bias',shape=[cnum])
    return  tf.nn.leaky_relu(x+bias)`
`def build_SNGAN_discriminator(x,mask,batch_size,reuse=False,training=True):
	with tf.variable_scope('discriminator',reuse=reuse):
		cnum=64
		x=dis_conv(x,cnum,name='conv1',training=training)
		x=dis_conv(x,2*cnum,name='conv2',training=training)
		x=dis_conv(x,4*cnum,name='conv3',training=training)
		x=dis_conv(x,4*cnum,name='conv4',training=training)
		x=dis_conv(x,4*cnum,name='conv5',training=training)
                return x
`

**Comment by JiahuiYu** [comment]
@shihaobai I have six discriminator convolutions in build_SNGAN_discriminator.

**Comment by nogu-atsu** [comment]
I implemented DeepFillv2 and now training it, but the quality of the generative results aren't good.

How many iterations did you need to get the high-quality results in the paper?
I'm training the model for 200,000 iterations with minibatch size 16. Is it enough?

**Comment by annzheng** [comment]
@JiahuiYu  Hi, thanks you code and paper :)
I try to train the DeepFillv2 model now. Since there are too many parameters, the training speed is very slow. But I don't know how to slim the model width to make training more efficient. Should I reduce the number of conv layer or the  channel number? Thanks in advance!

**Comment by xiaosean** [comment]
@JiahuiYu Hi, thank your code and paper
I have a question about the DeepFillv2 discriminator,
Are all layer in discriminator apply like below

`x = leaky_relu(spectral_norm(x))`

Including the first and last layer output?

**Comment by JiahuiYu** [comment]
@xiaosean Yes. `x = leaky_relu(conv2d_spectral_norm(x))`.

**Comment by JiahuiYu** [comment]
@aiueogawa Yes. The discriminator remain conv2d. Though I think it may be interesting to see whether gated conv2d help training inpainting network or not. Intuitively I guess a discriminator with gated conv2d may have better discriminative representation.

**Comment by aiueogawa** [comment]
@JiahuiYu 
Thank you for your response! 
I also think discriminator with gated convolutions would work better for gated convolutions could pay attention to the part around the mask and user inputs. I'll try it later.

I have three more questions.
1. What learning rate and optimizer did you use?
2. In the following mention, how did you do that for free-form mask?
>For efficiency, we can also restrict the search range of contextual attention module from the whole image to a local neighborhood, which can make the run-time significantly faster while maintaining overall quality of the results.
3. How did you incorporate user inputs in implementation? Did you put them into `mask` ?

**Comment by theahura** [comment]
Hey Jiahui, 

Thanks for being so responsive on this thread about deepfillv2. 
Quick question: you mention in the paper that your model has 4.1M parameters. Is that just the generator model or does that include the discriminator too?

**Comment by JiahuiYu** [comment]
@theahura After training, we only use the generator for inpainting. Thus 4.1M is only the generator.

**Comment by theahura** [comment]
A quick follow up. In the paper you describe using a coarse network and a refinement network from your previous work. Is the coarse network trained separately/has its own loss as per the previous work? My own implementations of deepfillv2 have trouble learning anything useful in the coarse network. 

Thanks again for your help and quick answers!

**Comment by JiahuiYu** [comment]
@theahura The coarse network exactly follows the training of this code, where we use a pixel-wise loss as additional supervision. They are trained jointly, instead of separately.

**Comment by JiahuiYu** [comment]
@theahura Yes.

**Comment by aiueogawa** [comment]
@JiahuiYu 
I'm a bit confused about the **activation function at the last layer in the discriminator** in DeepFillv2.

You said two different things about it.
1. you used the same setting of official SN-GANs where no activation (identity activation) is used in https://github.com/JiahuiYu/generative_inpainting/issues/62#issuecomment-402217672.
2. you used leaky relu in https://github.com/JiahuiYu/generative_inpainting/issues/62#issuecomment-421848081.
I wonder which is right.

Intuitively, SN-PatchGAN loss wants output of the discriminator for fake images generated from the generator to be -1 while that for real images to be +1, hence the activation at the last layer is desired to be symmetric in its domain.
This idea leads me to no activation (identity activation).
Tanh function seems a good alternative which is not only symmetric but also seems 1-Lipschitz and lies in (-1, 1).
Considering `Ex∼Pdata[ReLU(1−D(x))]+ Ez∼Pz[ReLU(1+D(G(z)))]`, I think `D(x)` should lie in `(-1, 1)`.

What activation function did you use as that in the last layer in the discriminator for DeepFillv2?
And how do you think of it?

**Comment by JiahuiYu** [comment]
@aiueogawa Hi thanks for your interest and good question.

Both my answers of #62 is correct. We use the **loss function** of SN-GANs as claimed in 
https://github.com/JiahuiYu/generative_inpainting/issues/62#issuecomment-402217672 (Please do not assume layers and activation functions to be the same.), and we do use leaky relu for all convolutional layers in discriminator as claimed in 
https://github.com/JiahuiYu/generative_inpainting/issues/62#issuecomment-421848081 (All layers including last layer).

What you are considering is about the activation function of last layer. You are correct that intuitively activation at the last layer is desired to be symmetric in its domain. However, I believe as long as the activation function can cover the domain of (-1, 1) and 1-Lipschitz condition holds, it is fine. Practically we find leaky relu (piecewise linear) works well. I guess using other activation function in last layer may not affect performance a lot. Feel free to report your results here if you have a try.

**Comment by aiueogawa** [comment]
@JiahuiYu Thank you! I see.
As you say, leaky ReLU also seems good and is.
In terms of the magnitude of gradients, leaky ReLU seems better than tanh, gradients of which might vanish.

**Comment by khemrajrathore** [comment]
@JiahuiYu I implemented the algorithm for generating random mask as suggested in the paper. Following are the masks which the algorithm generates. 
The value of hyperparameter are 
imageHeight, imageWidth = 256, 256
maxVertex = 20
maxLength = 200
maxAngle = 210
maxBrushWidth = 30
Actually, the result is not satisfactory, what could be the ideal hyperparameter values??

![demo](https://user-images.githubusercontent.com/25260675/46731573-ae8f5080-cca8-11e8-99b4-45eb33a8c881.png)
![demo1](https://user-images.githubusercontent.com/25260675/46731574-ae8f5080-cca8-11e8-92c1-0aaacbfebec2.png)
![demo2](https://user-images.githubusercontent.com/25260675/46731576-af27e700-cca8-11e8-9b51-f84010fb5065.png)
![demo4](https://user-images.githubusercontent.com/25260675/46731577-af27e700-cca8-11e8-834b-f8233c28680f.png)

**Comment by JiahuiYu** [comment]
@khemrajrathore The joints between two lines should be very smooth as shown in our paper. It is done by setting the joint width the same as line width.
The original paper has a typo. Please take a reference of [this algorithm](https://user-images.githubusercontent.com/22609465/46747759-b97dbd00-cc77-11e8-9b31-91590cb79f80.png)

ps: to make this issue compact, I have modified your image as links instead of direct display.

**Comment by aiueogawa** [comment]
@JiahuiYu 
What hyper parameters, maxVertex, maxLength, maxBrushWidth and maxAngle, did you use for free-form mask generation especially in CelebA-HQ?
And how many strokes in a free-form mask did you use?

**Comment by JiahuiYu** [comment]
@aiueogawa @khemrajrathore I am using:

```py
    min_num_vertex = 4
    max_num_vertex = 12
    mean_angle = 2*math.pi / 5
    angle_range = 2*math.pi / 15
    min_width = 12
    max_width = 40
```

**Comment by aiueogawa** [comment]
@JiahuiYu Thanks for your answer.
`min_num_vertex`, `mean_angle`, `angle_range`, `min_width` are not in your free-form mask generation algorithm.
I expect them to be as follows:

- `num_vertex` is chosen from the uniform distribution between `min_num_vertex` and `max_num_vertex`
- `angle` is chosen from the uniform distribution between `mean_angle - angle_range` and `mean_angle + angle_range`
- `width` is chosen from the uniform distribution between `min_width` and `max_width`

However, how `length` is chosen?
Furthermore, how `num_strokes` is chosen?

**Comment by khemrajrathore** [comment]
@JiahuiYu @aiueogawa Thanks for the suggestion. With the value of hyperparameters discussed above these are the results 
min_num_vertex = 4
    max_num_vertex = 12
    mean_angle = 2*math.pi / 5
    angle_range = 2*math.pi / 15
    min_width = 12
    max_width = 40
[Demo0](https://user-images.githubusercontent.com/25260675/46804776-24ff8180-cd81-11e8-8052-f5f4e42eefcf.png) [Demo1](https://user-images.githubusercontent.com/25260675/46804778-24ff8180-cd81-11e8-9dc1-96e79e1bc0af.png) [Demo2](https://user-images.githubusercontent.com/25260675/46804779-25981800-cd81-11e8-9db9-3a1600dc9e2f.png) [Demo3](https://user-images.githubusercontent.com/25260675/46804780-25981800-cd81-11e8-804a-756e69893c12.png)

**Comment by aiueogawa** [comment]
@khemrajrathore 
Your code still involves bugs around drawing a circle at a joint point, e.g. your joints are too big, as @JiahuiYu said.
The radius of a circle should be equal to the half width of a line.

**Comment by JiahuiYu** [comment]
@aiueogawa Sorry I missed the information of average length. I set it as `average_length = math.sqrt(H*H+W*W) / 8. For more detail, I have updated code above.

**Comment by aiueogawa** [comment]
@JiahuiYu Thanks for revised information.
Your discriminator of SN-PatchGAN produces HxWxC outputs and the loss defined in the following definition is also the same size HxWxC, because ReLU is an element wise operation and expectations are taken over samples.

<img width="336" alt="image" src="https://user-images.githubusercontent.com/30866185/46828447-332bbd00-cdd6-11e8-8209-ad71f9fd396e.png">

How are the loss values are summed up into a single total loss.
Simple summation or calculate average of them?
This difference (sum or average) affects the proportion of contributions of reconstruction loss and SN-PatchGAN loss to the final objective of the generator, because you mentioned,

> our final objective function for inpainting network is only composed of pixel-wise l1 reconstruction loss and SN-PatchGAN loss with default loss balancing hyper-parameter as 1 : 1.

in the DeepFillv2 paper.

**Comment by JiahuiYu** [comment]
@aiueogawa Reduce mean, as shown [here]( https://github.com/JiahuiYu/neuralgym/blob/c0b5863c6aaae74aaeb8e2231f71fdbe586d1305/neuralgym/ops/gan_ops.py#L50-L64).

**Comment by aiueogawa** [comment]
@JiahuiYu 
You describe the following two options for memory and computational efficiency of contextual attention layer in DeepFillv1.

1. extracting background patches with strides to reduce the number of filters
2. downscaling resolution of foreground inputs before convolution and upscaling attention map after propagation

and another option in DeepFillv2.

3. restricting the search range of contextual attention module from the whole image to a local neighborhood

In DeepFillv2, what values did you use as `stride` and `downscale_rate`?
And did you restrict the search range in CelebA-HQ training without user-guidance channels.

**Comment by JiahuiYu** [comment]
Option 3 is not used in CelebA-HQ case. All results in paper is without option 3. The option 3 is for real user cases when images are very large.

**Comment by aiueogawa** [comment]
@JiahuiYu
OK, then in DeepFillv2, what values did you use as `stride` and `downscale_rate` for contextual attention?

**Comment by aiueogawa** [comment]
@JiahuiYu 
Q1:
How many times did you iterate training of a discriminator at a training of a generator?

Q2:
In DeepFillv2 paper,

>The overall mask generation algorithm is illustrated in Algorithm 1. Additionally we can sample multiple strokes in single image to mask multiple regions.

Are multiple strokes used in training?

**Comment by JiahuiYu** [comment]
@aiueogawa Please find the answer above.

**Comment by aiueogawa** [comment]
@JiahuiYu I read all answers above but there are no answers of both of my questions.

**Comment by JiahuiYu** [comment]
@aiueogawa I have opened a specific issue for your case. I do not understand what is confusing to you so we can communicate in issue #158. I am trying my best to help.

**Comment by khemrajrathore** [comment]
hello! On running the following line of code,
` x = x*mask + xin*(1.-mask)`
I am getting an error:
Dimensions must be equal, but are 3 and 256 for 'inpaint_net/mul_15' (op: 'Mul') with input shapes: [16,3,256,2], [1,256,256,1].
however, my generative layers are the same as yours. 
I didn't understand why 
` x = gen_conv(x, 3, 3,1, activation=None, name='conv17')` works for 3 cnum value too.
